File: ueueo-framework/ueueo-eventbus/src/main/java/com/ueueo/eventbus/distributed/InboxProcessManager.cs
﻿using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Options;
using Volo.Abp.BackgroundWorkers;
using Volo.Abp.EventBus.Distributed;

namespace Volo.Abp.EventBus.Boxes;

public class InboxProcessManager : IBackgroundWorker
{
    protected AbpDistributedEventBusOptions Options;//  { get; }
    protected IServiceProvider ServiceProvider;//  { get; }
    protected List<IInboxProcessor> Processors;//  { get; }

    public InboxProcessManager(
        IOptions<AbpDistributedEventBusOptions> options,
        IServiceProvider serviceProvider)
    {
        ServiceProvider = serviceProvider;
        Options = options.Value;
        Processors = new List<IInboxProcessor>();
    }

    public void StartAsync(CancellationToken cancellationToken = default)
    {
        for (var inboxConfig in Options.Inboxes.Values)
        {
            if (inboxConfig.IsProcessingEnabled)
            {
                var processor = ServiceProvider.GetRequiredService<IInboxProcessor>();
                processor.StartAsync(inboxConfig, cancellationToken);
                Processors.Add(processor);
            }
        }
    }

    public void StopAsync(CancellationToken cancellationToken = default)
    {
        for (var processor in Processors)
        {
            processor.StopAsync(cancellationToken);
        }
    }
}

File: ueueo-frameworks/ueueo-eventbus/src/main/java/com/ueueo/eventbus/distributed/InboxProcessManager.java
package com.ueueo.eventbus.distributed;

import com.ueueo.backgroundworkers.IBackgroundWorker;
import lombok.Getter;

import java.util.ArrayList;
import java.util.List;

@Getter
public class InboxProcessManager implements IBackgroundWorker {
    protected AbpDistributedEventBusOptions options;
    protected IInboxProcessor inboxProcessor;
    protected List<IInboxProcessor> processors;

    public InboxProcessManager(
            AbpDistributedEventBusOptions options,
            IInboxProcessor inboxProcessor) {
        this.inboxProcessor = inboxProcessor;
        this.options = options;
        this.processors = new ArrayList<>();
    }

    @Override
    public void start() {
        for (InboxConfig inboxConfig : options.getInboxes().values()) {
            if (inboxConfig.isProcessingEnabled()) {
                inboxProcessor.start(inboxConfig);
                processors.add(inboxProcessor);
            }
        }
    }

    @Override
    public void stop() {
        for (IInboxProcessor processor : processors) {
            processor.stop();
        }
    }
}
